### PR TITLE
new command api

### PR DESCRIPTION
**Pull request by portlek**

### CODE REVIEW

File: src/main/java/club/issizler/okyanus/api/Mod.java [hunk 1]
```diff
@@ -33,16 +33,16 @@ public Server getServer() {
...
     /**
-     * Registers a {@link CommandBuilder}
+     * Registers a {@link ICommand}
      *
-     * @param builder The command to register
+     * @param command The command to register
      */
-    public void registerCommand(CommandBuilder builder) {
-        getServer().registerCommand(builder);
+    public void registerCommand(ICommand command) {
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/Mod.java]: How would registering the old builder-based commands work if you are just using a single method for registration. Shouldn't you have a deprecated compatibility method for that?

ps: You might want to test your changes against a mod like https://github.com/okyanus-mods/OKAuth that contains old builder-based commands.

File: src/main/java/club/issizler/okyanus/api/Server.java [hunk 2]
```diff
@@ -58,30 +57,6 @@
      */
     boolean isMainThread();
 
-    /**
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/Server.java]: You're removing these, but what about the mods that use them? Shouldn't you just deprecate them instead?

File: src/main/java/club/issizler/okyanus/api/Server.java [hunk 3]
```diff
@@ -91,10 +66,18 @@
...
     /**
-     * Execute a console command
+     * Runs the given command and sends it to the command
+     * dispatcher from the implementing dispatcher.
      *
-     * @param command Command to execute
+     * @param command the command to handle, without the "/"
      */
-     void exec(String command);
+    void runCommand(String command);
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/Server.java:79]: Again, if you are going to rename this, keep the old, but deprecate it instead.

File: src/main/java/club/issizler/okyanus/api/cmd/CommandRunnable.java [hunk 4]
```diff
@@ -1,16 +1,9 @@
...
 
-/**
- * Your should either
- * - Implement this in a class
- * - Create an anonymous class implementing this
- * <p>
- * And pass it to {@link CommandBuilder#run(CommandRunnable)} to make your
- * command actually do something
- */
+@Deprecated
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmd/CommandRunnable.java:8]: You're deprecating CommandRunnable, but what alternative are you providing? Even in runtime, you seem to keep using it.

File: src/main/java/club/issizler/okyanus/api/cmd/CommandSource.java [hunk 5]
```diff
@@ -5,9 +5,7 @@
 
 import java.util.Optional;
 
-/**
- * This is the source of a command
- */
+@Deprecated
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmd/CommandSource.java:11]: Same as `CommandRunnable`. Deprecated without an alternative.

File: src/main/java/club/issizler/okyanus/api/cmdnew/ArgumentType.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package club.issizler.okyanus.api.cmdnew;
+
+public enum ArgumentType {
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmdnew/ArgumentType.java]: If you're keeping this class, shouldn't you just... not deprecate the one in the old package?

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public String getLabel() {
+        for (CommandOption option : commandOptions) {
+            if (option.type() != OptionType.LABEL) continue;
+
+            return (String) option.value();
+        }
+
+        return "";
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java:36]: Instead of returning an empty string, wouldn't it better to make it optional or something?

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public CommandRunnable getRunnable() {
+        for (CommandOption option : commandOptions) {
+            if (option.type() != OptionType.RUN) continue;
+
+            return (CommandRunnable) option.value();
+        }
+
+        return new MckCommandRunnable();
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java:69]: Same as `getLabel`. Just make it optional if you're going to return a mock object.

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandOption.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+     * Gets option of value
+     * @return {@link OptionType}
+     */
+    OptionType type();
+
+    /**
+     * Gets value of option
+     * @return object
+     */
+    Object value();
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmdnew/CommandOption.java]: Assuming this returns `X`, can't you just make it return `X` instead of `Object`?

File: src/main/java/club/issizler/okyanus/api/cmd/CommandRunnable.java (hunk 4 above)
Comment (Admicos) [reply src/main/java/club/issizler/okyanus/api/cmd/CommandRunnable.java:8]: I see that you just moved the class. You should keep using this one instead of having duplicate code, as we'll eventually merge both `cmd` and `cmdnew` into a single package on 1.0.0.

File: src/main/java/club/issizler/okyanus/api/cmdnew/ICommand.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
+package club.issizler.okyanus.api.cmdnew;
+
+import java.util.List;
+
+public interface ICommand {
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/cmdnew/ICommand.java]: I never use the `I` prefix when referring to interfaces. You probably shouldn't as well. I feel like it's kind of unneccesary. We don't go around having variables like `publicStringName`, or classes like `FooClass` do we?

File: src/main/java/club/issizler/okyanus/api/perms/PermissionService.java [hunk 11]
```diff
@@ -38,4 +38,22 @@
...
     void removePermission(Permissible permissible, String perm);
 
+    /**
+     * Sets whether or not the permissible is an operator,
+     * or otherwise holds all permissions.
+     *
+     * @param op {@code true} to set to an operator,
+     * {@code false} to set to a non-operator.
+     */
+    void setOp(Permissible permissible, boolean op);
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/perms/PermissionService.java]: I am not entirely sure if we should merge OP-ness with permissions, as the OP permission is sort of "internal" to the game. and permission services would need to access MC internals to get if the player is actually OP or not.

File: src/main/java/club/issizler/okyanus/api/perms/Permissible.java [hunk 12]
```diff
@@ -35,16 +35,4 @@ default void removePermission(String perm) {
...
-     * Gets this permissible's identifier
-     * <p>
-     * It can be:
-     * * CONSOLE for the server console
-     * * Player UUID for players
-     * * UNKNOWN if something's bugged
-     *
-     * @return This permissible's identifier
-     */
-    String getIdentifier();
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/perms/Permissible.java]: What would be the alternative to this identifier, and wouldn't removing it right now break the API?

**Review by Admicos — CHANGES_REQUESTED** [review]
While the end result is really good, and the Runtime PR was great, this PR needs some improvement. 

You seem to be breaking the old API, which you shouldn't. And most new classes don't have much documentation, which I'd like to see.

Other than those (and anything I said on the specific comments), this could get merged.

PS: Can you change your branch target from `master` to `develop`? I stopped working on master after releasing 0.1.0

File: src/main/java/club/issizler/okyanus/api/Mod.java (hunk 1 above)
Comment (portlek) [reply src/main/java/club/issizler/okyanus/api/Mod.java]: yes i didn't add it, i'll

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandOption.java (hunk 9 above)
Comment (portlek) [reply src/main/java/club/issizler/okyanus/api/cmdnew/CommandOption.java]: yes i forgeted it

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java (hunk 7 above)
Comment (portlek) [reply src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java:36]: i don't want to use nullable objects/things

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java (hunk 8 above)
Comment (portlek) [reply src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java:69]: i don't want to use nullable objects/things

File: src/main/java/club/issizler/okyanus/api/Mod.java [hunk 13]
```diff
@@ -33,16 +34,25 @@ public Server getServer() {
...
 
     /**
      * Registers a {@link CommandBuilder}
      *
-     * @param builder The command to register
+     * @param command The command to register
      */
-    public void registerCommand(CommandBuilder builder) {
-        getServer().registerCommand(builder);
+    public void registerCommand(CommandBuilder command) {
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/Mod.java:56]: You might want to mark this as deprecated

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java (hunk 7 above)
Comment (Admicos) [reply src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java:36]: Optional isn't nullable. You have to explicitly check optional.isPresent if you are using Optionals

File: src/main/java/club/issizler/okyanus/api/entity/Player.java [hunk 14]
```diff
@@ -18,7 +16,7 @@
      * @param returnFluids Should we return fluids?
      * @return Block, or empty if we couldn't find anything or we found an entity
      */
-    Optional<Block> getTargetBlock(double distance, boolean returnFluids);
+    Block getTargetBlock(double distance, boolean returnFluids);
```
Comment (Admicos) [inline src/main/java/club/issizler/okyanus/api/entity/Player.java]: This breaks the API. Also please make this into another PR if you are planning to remove these optionals.

**Review by Admicos — CHANGES_REQUESTED** [review]
Definitely better than your last submission, but some changes are still needed IMO.

File: src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java (hunk 7 above)
Comment (portlek) [reply src/main/java/club/issizler/okyanus/api/cmdnew/CommandEnvelope.java:36]: i mean it's using nulls

**Review by Admicos — APPROVED** [review]
Alright, this should be enough, I hope.

I'll clean anything else up if I see something
